File: leetcode/trapping-rain-water-ii.cpp

#include<iostream>
#include<algorithm>
#include<vector>
#include<queue>

#define INT int
using namespace std;
// g++ -ggdb -std=c++1y trapping-rain-water-ii.cpp -o trapping-rain-water-ii.exe
class tupleItem{
private:
    INT x = 0;
    INT y = 0;
    INT h = 0;
public:
    tupleItem(INT xx, INT yy, INT hh):x(xx),y(yy),h(hh){}
    INT get_x () const { return x;}
    INT get_y () const { return y;}
    INT get_h () const { return h;}
    /*
    bool operator > (const tupleItem &rhs) const{
        return h > rhs.h;
    }
    */
};



class Solution {
public:
    struct compareGreater{
        bool operator()(const tupleItem &lhs, const tupleItem &rhs){
            return (lhs.get_h() > rhs.get_h());
        }
    };

    INT trapRainWater(vector<vector<int>>& heightMap) {
        INT rlen = heightMap.size();
        INT clen = 0;
        INT res = 0; 
        if(rlen>0){
            clen = heightMap[0].size();
        }
        else return 0;
        // marked the visited map
        vector<vector<bool>> visited;
        visited.resize(rlen);
        //for(auto vi : visited){
        for(INT i = 0; i < rlen; i++)
            visited[i].resize(clen, false);
        //}
        // organize data
        int r_i = 0;
        //vector<tupleItem> vtmp;
        std::priority_queue<tupleItem, vector<tupleItem>, compareGreater> vtmp;

        for(int i = 0 ; i < clen; i++){ 
             tupleItem tmp0(0, i, heightMap[0][i]);
             vtmp.push(tmp0);
             visited[0][i]=true;
             tupleItem tmp1(rlen-1, i, heightMap[rlen-1][i]);
             vtmp.push(tmp1);
             visited[rlen-1][i]=true;
        }
        for(int i = 1 ; i < rlen-1; i++){ 
             tupleItem tmp0(i, 0, heightMap[i][0]);
             vtmp.push(tmp0);
             visited[i][0]=true;
             tupleItem tmp1(i, clen-1, heightMap[i][clen-1]);
             vtmp.push(tmp1);
             visited[i][clen-1]=true;
        }
       
        //std::make_heap(vtmp.begin(), vtmp.end());
        // heap
        //cout<<vtmp.size()<<endl;
        vector<std::pair<INT, INT>> direction = {{0,1},{0,-1},{1,0},{-1,0}};
        INT max_h = -1;
        INT store = 0;
        while(!vtmp.empty()){
            tupleItem tmp = vtmp.top(); 
            vtmp.pop();
            INT h = tmp.get_h();
            INT x = tmp.get_x();
            INT y = tmp.get_y();
            //cout<<h<<endl;
            if(max_h < h) {
                max_h = h;
            }
            //cout<<max_h<<endl;
            //store += (max_h - h);
            for( auto it : direction){
                INT xx = x + it.first;
                INT yy = y + it.second;
                if( !(xx < rlen && xx >= 0 && yy < clen && yy >= 0 ) || visited[xx][yy]) continue;
                if( heightMap[xx][yy] < max_h ) {
                    //cout<<"hdelta "<<max_h - heightMap[xx][yy]<<" store = "<<store<<endl;
                    store+=(max_h - heightMap[xx][yy]);
                }
                //cout<<"push "<<xx<<", "<<yy<<endl;
                visited[xx][yy] = true; 
                tupleItem tmp(xx, yy, heightMap[xx][yy]);
                vtmp.push(tmp);
            }
        }
        return store;
    }
};


int main(){
    Solution a;
    //vector<vector<int>> hm = {{12,13,1,12},{13,4,13,12},{13,8,10,12},{12,13,12,12},{13,13,13,13}};
    vector<vector<int>> hm = {{5,5,5,1},{5,1,1,5},{5,1,5,5},{5,2,5,8}};
    cout<<a.trapRainWater(hm)<<endl;
    return 0;
}

File: leetcode/py/trapping-rain-water-ii.py

import heapq

class tupleItem(object):
    def __init__(self, x, y, h):
        self._r = x
        self._c = y
        self._h = h

    def __lt__(self, other):
        return self._h < other.get_h()
    def get_r(self):
        return self._r
    def get_c(self):
        return self._c
    def get_h(self):
        return self._h

class Solution(object): 
    def trapRainWater(self, heightMap):
        """
        :type heightMap: List[List[int]]
        :rtype: int
        """
        hp = []
        row = len(heightMap)
        if row <= 1:
            return 0
        col = len(heightMap[0])
        if col <= 1:
            return 0
        
        visited = []
        for r in range(0, row):
            visited.append([])
            for c in range(0, col):
                visited[r].append(False)
        for r in range(0, row):
            ob = tupleItem(r, col-1, heightMap[r][col-1] )
            visited[r][col-1] = True
            heapq.heappush(hp, ob)
            ob = tupleItem(r, 0, heightMap[r][0] )
            visited[r][0] = True
            heapq.heappush(hp, ob)
        for c in range(0, col):
            ob = tupleItem(0, c, heightMap[0][c] )
            heapq.heappush(hp, ob)
            visited[0][c] = True
            ob = tupleItem(row-1, c, heightMap[row-1][c])
            heapq.heappush(hp, ob)   
            visited[row-1][c] = True
        

        res = 0
        _max = 0
        direction = [[0, -1], [0, 1], [1, 0], [-1, 0]]
        while hp:
            t = heapq.heappop(hp)
            r, c, h = t.get_r(), t.get_c(), t.get_h()
            _max = max(h, _max)
            def in_range(x, y, row, col):
                if x >= 0 and x < row and y >= 0 and y < col:
                    return True
                return False
            for (rd, cd) in direction:
                x = r + rd
                y = c + cd
                if in_range(x, y, row, col) and visited[x][y] == False:
                    visited[x][y] = True
                    if heightMap[x][y] < _max:
                        res += _max - heightMap[x][y]
                    ob = tupleItem(x, y, heightMap[x][y])
                    heapq.heappush(hp, ob)
        return res

a = Solution()
matrix = [[1,4,3,1,3,2],[3,2,1,3,2,4],[2,3,3,2,3,1]]
matrix = [[12,13,1,12],[13,4,13,12],[13,8,10,12],[12,13,12,12],[13,13,13,13]]
print(a.trapRainWater(matrix))